common/flatpak-installation.c

```diff
@@ -991,6 +991,85 @@ _ostree_collection_ref_free0 (OstreeCollectionRef *ref)
   ostree_collection_ref_free (ref);
 }
 
+static gboolean
+flatpak_installation_check_installed_ref_missing_related_ref (FlatpakInstallation *self,
+                                                              const gchar         *remote_name,
+                                                              gchar               *full_ref,
+                                                              GCancellable        *cancellable)
+{
+  g_autoptr(FlatpakDir) dir  = NULL;
+  g_autoptr (GPtrArray) installed_related_refs = NULL;
+  g_autoptr (GPtrArray) remote_related_refs = NULL;
+  g_autoptr (GError) local_error = NULL;
+  guint j, k;
+
+  dir = flatpak_installation_get_dir (self, &local_error);
+  if (dir == NULL)
+    return FALSE;
+
+  installed_related_refs = flatpak_installation_list_installed_related_refs_sync (self,
+                                                                                  remote_name,
+                                                                                  full_ref,
+                                                                                  cancellable,
+                                                                                  &local_error);
+  if (local_error != NULL)
+    {
+      g_warning ("Unable to get installed related refs for %s: %s", full_ref, local_error->message);
+      return FALSE;
+    }
+
+  g_return_val_if_fail (installed_related_refs != NULL, FALSE);
+
+  remote_related_refs = flatpak_installation_list_remote_related_refs_sync (self,
+                                                                            remote_name,
+                                                                            full_ref,
+                                                                            cancellable,
+                                                                            &local_error);
+  if (local_error != NULL)
+    {
+      g_warning ("Unable to get remote related refs for %s: %s", full_ref, local_error->message);
+      return FALSE;
+    }
+
+  if (remote_related_refs == NULL)
+    return FALSE;
+
+  for (j = 0; j < remote_related_refs->len; j++)
+    {
+      FlatpakRelatedRef *remote_related_ref = g_ptr_array_index (remote_related_refs, j);
+      g_autofree gchar *remote_related_full_ref = flatpak_ref_format_ref (FLATPAK_REF (remote_related_ref));
+      gboolean should_download = flatpak_related_ref_should_download (remote_related_ref);
+      gboolean related_ref_found = FALSE;
+
+      if (flatpak_dir_ref_is_masked (dir, remote_related_full_ref))
+          continue;
+
+      if (should_download)
+        {
+          for (k = 0; k < installed_related_refs->len; k++)
```

Comment:
I think this is doing quite a bit of unnecessary work by finding all matching installed extensions and then comparing to them. A better approach would be to just get the remote related refs, and then call `flatpak_dir_get_if_deployed (dir, remote_related_full_ref, NULL, NULL)` to see if its installed already.